src/codegen/compile.ml

```diff
@@ -272,20 +273,17 @@ module Const = struct
   type v =
     | Fun of (unit -> int32) * fun_rhs (* function pointer calculated upon first use *)
     | Message of int32 (* anonymous message, only temporary *)
-    | Obj of (string * t) list
+    | Obj of (string * v) list
     | Unit
-    | Array of t list (* also tuples, but not nullary *)
-    | Tag of (string * t)
-    | Opt of t
+    | Array of v list (* also tuples, but not nullary *)
+    | Tag of (string * v)
+    | Opt of v
     | Lit of lit
 
-  (* A constant known value together with a vanilla pointer.
-     Typically a static memory location, could be an unboxed scalar.
-     Filled on demand.
-   *)
-  and t = (int32 Lib.Promise.t * v)
-
-  let t_of_v v = (Lib.Promise.make (), v)
+  let eq v1 v2 = match v1, v2 with
+    | Lit l1, Lit l2 -> lit_eq l1 l2
+    | Fun (f1_fp, _), Fun (f2_fp, _) -> f1_fp() = f2_fp()
+    | _ -> v1 = v2
```

Comment:
This  definition of `eq` make me nervous because it will just use Ocaml equality for nested `Fun` and `Lit` constants.  It's probably better to just define it explicitly on all cases....